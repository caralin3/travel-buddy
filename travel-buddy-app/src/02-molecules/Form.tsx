import React, { FormEvent } from 'react';
import { Link } from 'react-router-dom';
import { Button, Form as BSForm, FormGroup as BSFormGroup, FormGroupProps, Spinner } from 'reactstrap';
import { LinkItem } from '../types';

export interface FormProps {
  additionalLinks?: LinkItem[];
  loading?: boolean;
  onSubmit: () => void;
  title?: string;
}

export const Form: React.FC<FormProps> = ({ additionalLinks, children, loading = false, onSubmit, title }) => {
  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    onSubmit();
  };

  return (
    <BSForm
      className="d-flex flex-column align-items-center shadow-sm px-5 py-4 border rounded"
      onSubmit={handleSubmit}
    >
      {!!title && <h2 className="mb-4">{title}</h2>}
      {children}
      {!loading ? (
        <Button color="primary" className="mt-3 text-align-center align-self-center">
          Submit
        </Button>
      ) : (
        <Button className="mt-3 px-4 text-align-center align-self-center" disabled={true}>
          <Spinner size="sm">Loading...</Spinner>
        </Button>
      )}
      {!!additionalLinks &&
        additionalLinks.length > 0 &&
        additionalLinks.map((link) => (
          <Link className="mt-3" key={link.path} to={link.path}>
            {link.label}
          </Link>
        ))}
    </BSForm>
  );
};

export const FormGroup: React.FC<FormGroupProps> = ({ children, props }) => (
  <BSFormGroup className="d-flex flex-column align-items-start" {...props}>
    {children}
  </BSFormGroup>
);
