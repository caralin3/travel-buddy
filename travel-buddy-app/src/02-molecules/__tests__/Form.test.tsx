import { fireEvent, render } from '@testing-library/react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CostInput, Form, FormGroup, FormProps, Label, Select } from '../Form';

describe('Form', () => {
  describe('Form component', () => {
    const props: FormProps = {
      additionalLinks: [{ path: '#', label: 'Link' }],
      onSubmit: jest.fn(),
      title: 'Form',
    };

    it('should render correctly', () => {
      const wrapper = shallow(<Form {...props} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render loading correctly', () => {
      const wrapper = shallow(<Form {...props} loading={true} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should handle submit correctly', async () => {
      const submitMock = jest.fn();
      const { container } = render(<Form onSubmit={submitMock} />);
      const form = container.querySelector('form');
      if (form) {
        fireEvent.submit(form);
        expect(submitMock).toHaveBeenCalled();
      }
    });
  });
  describe('FormGroup component', () => {
    it('should render correctly', () => {
      const wrapper = shallow(<FormGroup></FormGroup>);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
  describe('Label component', () => {
    it('should render correctly', () => {
      const wrapper = shallow(<Label>test label</Label>);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render required correctly', () => {
      const wrapper = shallow(<Label required>test label</Label>);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
  describe('Select component', () => {
    it('should render correctly', () => {
      const selectMock = jest.fn();
      const wrapper = shallow(
        <Select
          id="select2"
          label="Label2"
          onSelect={selectMock}
          options={[
            { label: 'option1', value: '1' },
            { label: 'option2', value: '2' },
          ]}
          value="2"
        />
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render required correctly', () => {
      const selectMock = jest.fn();
      const wrapper = shallow(
        <Select
          required
          id="select"
          label="Label"
          onSelect={selectMock}
          options={[
            { label: 'option1', value: '1' },
            { label: 'option2', value: '2' },
          ]}
          value="1"
        />
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
  describe('CostInput component', () => {
    it('should render correctly', () => {
      const wrapper = shallow(<CostInput />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should render correctly', () => {
      const wrapper = shallow(<CostInput symbol="$" />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
