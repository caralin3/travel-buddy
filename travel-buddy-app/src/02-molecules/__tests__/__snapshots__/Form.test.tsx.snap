// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form CostInput component should render correctly 1`] = `
<InputGroup
  tag="div"
>
  <InputGroupText
    tag="span"
  >
    $
  </InputGroupText>
  <Input
    type="number"
  />
</InputGroup>
`;

exports[`Form CostInput component should render correctly 2`] = `
<InputGroup
  tag="div"
>
  <InputGroupText
    tag="span"
  >
    $
  </InputGroupText>
  <Input
    type="number"
  />
</InputGroup>
`;

exports[`Form Form component should render correctly 1`] = `
<Form
  className="d-flex flex-column shadow-sm px-5 py-4 border rounded"
  onSubmit={[Function]}
  tag="form"
>
  <h2
    className="mb-4"
  >
    Form
  </h2>
  <Button
    className="mt-3 text-align-center align-self-center"
    color="primary"
    tag="button"
  >
    Submit
  </Button>
  <Link
    className="mt-3"
    key="#"
    to="#"
  >
    Link
  </Link>
</Form>
`;

exports[`Form Form component should render loading correctly 1`] = `
<Form
  className="d-flex flex-column shadow-sm px-5 py-4 border rounded"
  onSubmit={[Function]}
  tag="form"
>
  <h2
    className="mb-4"
  >
    Form
  </h2>
  <Button
    className="mt-3 px-4 text-align-center align-self-center"
    color="secondary"
    disabled={true}
    tag="button"
  >
    <Spinner
      size="sm"
      tag="div"
      type="border"
    >
      Loading...
    </Spinner>
  </Button>
  <Link
    className="mt-3"
    key="#"
    to="#"
  >
    Link
  </Link>
</Form>
`;

exports[`Form FormGroup component should render correctly 1`] = `
<FormGroup
  className="d-flex flex-column align-items-start"
  tag="div"
/>
`;

exports[`Form Label component should render correctly 1`] = `
<Label
  tag="label"
  widths={
    Array [
      "xs",
      "sm",
      "md",
      "lg",
      "xl",
      "xxl",
    ]
  }
>
  test label
</Label>
`;

exports[`Form Label component should render required correctly 1`] = `
<Label
  tag="label"
  widths={
    Array [
      "xs",
      "sm",
      "md",
      "lg",
      "xl",
      "xxl",
    ]
  }
>
  test label
  <span
    className="text-danger"
  >
    *
  </span>
</Label>
`;

exports[`Form Select component should render correctly 1`] = `
<FormGroup>
  <Label
    for="select2"
    required={false}
  >
    Label2
  </Label>
  <Input
    defaultValue="2"
    id="select2"
    name="select"
    onChange={[MockFunction]}
    required={false}
    type="select"
  >
    <option
      key="1"
      value="1"
    >
      option1
    </option>
    <option
      key="2"
      value="2"
    >
      option2
    </option>
  </Input>
</FormGroup>
`;

exports[`Form Select component should render required correctly 1`] = `
<FormGroup>
  <Label
    for="select"
    required={true}
  >
    Label
  </Label>
  <Input
    defaultValue="1"
    id="select"
    name="select"
    onChange={[MockFunction]}
    required={true}
    type="select"
  >
    <option
      key="1"
      value="1"
    >
      option1
    </option>
    <option
      key="2"
      value="2"
    >
      option2
    </option>
  </Input>
</FormGroup>
`;
